// @sudoku/stores/strategy.js
import { writable } from 'svelte/store';

// 添加候选数管理的工具函数
export const CandidateManager = {
    // 获取单元格的候选数
    getCandidates(grid, row, col) {
        if (grid[row][col] !== 0) return [];
        const candidates = new Set([1, 2, 3, 4, 5, 6, 7, 8, 9]);
        
        // 检查行
        for (let c = 0; c < 9; c++) {
            if (grid[row][c] !== 0) candidates.delete(grid[row][c]);
        }
        
        // 检查列
        for (let r = 0; r < 9; r++) {
            if (grid[r][col] !== 0) candidates.delete(grid[r][col]);
        }
        
        // 检查3x3宫格
        const boxRow = Math.floor(row / 3) * 3;
        const boxCol = Math.floor(col / 3) * 3;
        for (let r = 0; r < 3; r++) {
            for (let c = 0; c < 3; c++) {
                if (grid[boxRow + r][boxCol + c] !== 0) {
                    candidates.delete(grid[boxRow + r][boxCol + c]);
                }
            }
        }
        
        return Array.from(candidates);
    },

    // 获取整个网格的候选数
    getGridCandidates(grid) {
        const candidates = Array(9).fill().map(() => 
            Array(9).fill().map(() => []));
        
        for (let row = 0; row < 9; row++) {
            for (let col = 0; col < 9; col++) {
                if (grid[row][col] === 0) {
                    candidates[row][col] = this.getCandidates(grid, row, col);
                }
            }
        }
        
        return candidates;
    }
};

// 添加X-Wing策略的实现
export const XWingStrategy = {
    // 查找行X-Wing模式
    findRowXWing(grid, candidatesMap) {
        const patterns = [];
        // 遍历所有可能的数字1-9
        for (let digit = 1; digit <= 9; digit++) {
            // 记录每行中数字出现的位置
            const rowPositions = Array(9).fill().map(() => []);
            
            // 遍历每一行，找出候选数digit出现的列位置
            for (let row = 0; row < 9; row++) {
                for (let col = 0; col < 9; col++) {
                    const key = `${col},${row}`;
                    if (candidatesMap[key] && candidatesMap[key].includes(digit)) {
                        rowPositions[row].push(col);
                    }
                }
            }
            
            // 查找具有相同两个位置的行
            for (let row1 = 0; row1 < 8; row1++) {
                // 确保第一行恰好有两个候选位置
                if (rowPositions[row1].length !== 2) continue;
                
                for (let row2 = row1 + 1; row2 < 9; row2++) {
                    // 确保第二行也恰好有两个候选位置，且位置与第一行相同
                    if (rowPositions[row2].length === 2 &&
                        rowPositions[row1][0] === rowPositions[row2][0] &&
                        rowPositions[row1][1] === rowPositions[row2][1]) {
                        
                        // 找到X-Wing模式，生成效果
                        const col1 = rowPositions[row1][0];
                        const col2 = rowPositions[row1][1];
                        const changes = [];
                        
                        // 检查其他行的这两列是否有相同的候选数，如果有则可以删除
                        for (let r = 0; r < 9; r++) {
                            if (r !== row1 && r !== row2) {
                                const key1 = `${col1},${r}`;
                                const key2 = `${col2},${r}`;
                                
                                if (candidatesMap[key1] && candidatesMap[key1].includes(digit)) {
                                    changes.push({
                                        type: 'highlight_candidate',
                                        row: r,
                                        col: col1,
                                        digit: digit,
                                        highlightType: 'eliminated'
                                    });
                                }
                                
                                if (candidatesMap[key2] && candidatesMap[key2].includes(digit)) {
                                    changes.push({
                                        type: 'highlight_candidate',
                                        row: r,
                                        col: col2,
                                        digit: digit,
                                        highlightType: 'eliminated'
                                    });
                                }
                            }
                        }
                        
                        // 高亮X-Wing的四个角
                        changes.push(
                            {
                                type: 'highlight_candidate',
                                row: row1,
                                col: col1,
                                digit: digit,
                                highlightType: 'xwing'
                            },
                            {
                                type: 'highlight_candidate',
                                row: row1,
                                col: col2,
                                digit: digit,
                                highlightType: 'xwing'
                            },
                            {
                                type: 'highlight_candidate',
                                row: row2,
                                col: col1,
                                digit: digit,
                                highlightType: 'xwing'
                            },
                            {
                                type: 'highlight_candidate',
                                row: row2,
                                col: col2,
                                digit: digit,
                                highlightType: 'xwing'
                            }
                        );
                        
                        if (changes.length > 4) { // 如果除了四个角外还有其他要删除的候选数
                            patterns.push({
                                digit,
                                changes
                            });
                        }
                    }
                }
            }
        }
        return patterns;
    },

    // 查找列X-Wing模式
    findColXWing(grid, candidatesMap) {
        const patterns = [];
        // 遍历所有可能的数字1-9
        for (let digit = 1; digit <= 9; digit++) {
            // 记录每列中数字出现的位置
            const colPositions = Array(9).fill().map(() => []);
            
            // 遍历每一列，找出候选数digit出现的行位置
            for (let col = 0; col < 9; col++) {
                for (let row = 0; row < 9; row++) {
                    const key = `${col},${row}`;
                    if (candidatesMap[key] && candidatesMap[key].includes(digit)) {
                        colPositions[col].push(row);
                    }
                }
            }
            
            // 查找具有相同两个位置的列
            for (let col1 = 0; col1 < 8; col1++) {
                // 确保第一列恰好有两个候选位置
                if (colPositions[col1].length !== 2) continue;
                
                for (let col2 = col1 + 1; col2 < 9; col2++) {
                    // 确保第二列也恰好有两个候选位置，且位置与第一列相同
                    if (colPositions[col2].length === 2 &&
                        colPositions[col1][0] === colPositions[col2][0] &&
                        colPositions[col1][1] === colPositions[col2][1]) {
                        
                        // 找到X-Wing模式，生成效果
                        const row1 = colPositions[col1][0];
                        const row2 = colPositions[col1][1];
                        const changes = [];
                        
                        // 检查其他列的这两行是否有相同的候选数，如果有则可以删除
                        for (let c = 0; c < 9; c++) {
                            if (c !== col1 && c !== col2) {
                                const key1 = `${c},${row1}`;
                                const key2 = `${c},${row2}`;
                                
                                if (candidatesMap[key1] && candidatesMap[key1].includes(digit)) {
                                    changes.push({
                                        type: 'highlight_candidate',
                                        row: row1,
                                        col: c,
                                        digit: digit,
                                        highlightType: 'eliminated'
                                    });
                                }
                                
                                if (candidatesMap[key2] && candidatesMap[key2].includes(digit)) {
                                    changes.push({
                                        type: 'highlight_candidate',
                                        row: row2,
                                        col: c,
                                        digit: digit,
                                        highlightType: 'eliminated'
                                    });
                                }
                            }
                        }
                        
                        // 高亮X-Wing的四个角
                        changes.push(
                            {
                                type: 'highlight_candidate',
                                row: row1,
                                col: col1,
                                digit: digit,
                                highlightType: 'xwing'
                            },
                            {
                                type: 'highlight_candidate',
                                row: row1,
                                col: col2,
                                digit: digit,
                                highlightType: 'xwing'
                            },
                            {
                                type: 'highlight_candidate',
                                row: row2,
                                col: col1,
                                digit: digit,
                                highlightType: 'xwing'
                            },
                            {
                                type: 'highlight_candidate',
                                row: row2,
                                col: col2,
                                digit: digit,
                                highlightType: 'xwing'
                            }
                        );
                        
                        if (changes.length > 4) { // 如果除了四个角外还有其他要删除的候选数
                            patterns.push({
                                digit,
                                changes
                            });
                        }
                    }
                }
            }
        }
        return patterns;
    }
};

// 数独策略枚举
export const STRATEGIES = {
    BASIC_ELIMINATION: {
        name: 'simple strategy',
        description: '检查同行、同列和同宫格中的数字，排除已经使用的数字',
        effect: (grid) => {
            console.log('使用基本消除策略');
            return [];
        },
    },
    X_WING: {
        name: 'X-Wing',
        description: '当两行（或两列）中，某个候选数仅出现在相同的两列（或两行）时，这四个单元格形成X-Wing结构。此时，可以排除其他行或列中相同位置的该候选数。',
        effect: (grid) => {
            // 获取当前的候选数
            const candidatesMap = {};
            for (let row = 0; row < 9; row++) {
                for (let col = 0; col < 9; col++) {
                    if (grid[row][col] === 0) {
                        const candidates = [];
                        for (let num = 1; num <= 9; num++) {
                            let canPlace = true;
                            // 检查行
                            for (let x = 0; x < 9; x++) {
                                if (grid[row][x] === num) {
                                    canPlace = false;
                                    break;
                                }
                            }
                            // 检查列
                            if (canPlace) {
                                for (let y = 0; y < 9; y++) {
                                    if (grid[y][col] === num) {
                                        canPlace = false;
                                        break;
                                    }
                                }
                            }
                            // 检查3x3方格
                            if (canPlace) {
                                const boxRow = Math.floor(row / 3) * 3;
                                const boxCol = Math.floor(col / 3) * 3;
                                for (let y = boxRow; y < boxRow + 3; y++) {
                                    for (let x = boxCol; x < boxCol + 3; x++) {
                                        if (grid[y][x] === num) {
                                            canPlace = false;
                                            break;
                                        }
                                    }
                                    if (!canPlace) break;
                                }
                            }
                            if (canPlace) {
                                candidates.push(num);
                            }
                        }
                        if (candidates.length > 0) {
                            candidatesMap[`${col},${row}`] = candidates;
                        }
                    }
                }
            }

            // 查找X-Wing模式
            const rowPatterns = XWingStrategy.findRowXWing(grid, candidatesMap);
            const colPatterns = XWingStrategy.findColXWing(grid, candidatesMap);
            
            // 返回找到的第一个有效模式
            const allPatterns = [...rowPatterns, ...colPatterns];
            if (allPatterns.length > 0) {
                return allPatterns[0].changes;
            }
            
            return [];
        }
    },
    LAST_REMAINING_1: {
        name: 'last remaining',
        description: '从未使用的数字中选择一个',
        effect: (grid) => {
            console.log('使用最后剩余策略');
            return [];
        },
    },
    SINGLE_CHOICE: {
        name: 'single choice',
        description: '填入唯一可能的数字',
        effect: (grid) => {
            console.log('使用单选策略');
            return [];
        },
    }
};

// 创建策略提示 store
function createStrategyHint() {
    const store = writable({
        strategy: STRATEGIES.BASIC_ELIMINATION,  // 设置默认策略
        position: null
    });

    return {
        subscribe: store.subscribe,
        setHint: (strategy, position) => {
            store.update(() => ({
                strategy: STRATEGIES[strategy] || STRATEGIES.BASIC_ELIMINATION,
                position
            }));
        },
        clear: () => {
            store.update(() => ({
                strategy: STRATEGIES.BASIC_ELIMINATION,  // 重置为默认策略
                position: null
            }));
        }
    };
}

export const strategyHint = createStrategyHint();
