// @sudoku/stores/strategy.js
import { writable } from 'svelte/store';

// 数独策略枚举
export const STRATEGIES = {
    BASIC_ELIMINATION: {
        name: 'simple strategy',
        description: '检查同行、同列和同宫格中的数字，排除已经使用的数字',
        effect:()=>{
            console.log('最后剩下一个策略');
        },
    },
    // 未来可以添加更多策略
    // TODO
    LAST_REMAINING_1: {
        name: 'last remaining',
        description: '从未使用的数字中选择一个',
        effect:()=>{
            console.log('最后剩下一个策略');
        },
    },
    SINGLE_CHOICE: {
        name: 'single choice',
        description: '填入唯一可能的数字',
        effect:()=>{
            console.log('单选策略');
        },
    }
};

// 创建策略提示 store
function createStrategyHint() {
    const { subscribe, set } = writable({
        strategy: null,
        position: null,
        value: null
    });

    return {
        subscribe,
        // 设置当前使用的策略和位置
        setHint: (strategyKey, position) => {
            console.log(strategyKey, position);
            // 在这里可以写入设置使用的策略和位置的逻辑
            
            set({
                strategy,
                position,
                value
            });
        },
        // 清除提示
        clear: () => {
            set({
                strategy: null,
                position: null,
                value: null
            });
        }
    };
}

export const strategyHint = createStrategyHint();
